Algorithms/Search/BinarySearcher.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using Algorithms.Sorting;
+
+namespace Algorithms.Search
+{
+    public class BinarySearcher<T> : IEnumerator<T>
+    {
+
+        private readonly IList<T> _collection;
+        private readonly Comparer<T> _comparer;
+
+        private int _currentItemIndex;
+        private int _leftIndex;
+        private int _rightIndex;
+
+        /// <summary>
+        /// The value of the current item
+        /// </summary>
+        public T Current {
+
+            get
+            {
+                return _collection[_currentItemIndex];
+            }
+
+        }
+
+
+        object IEnumerator.Current => Current;
+
+        /// <summary>
+        /// Class constructor
+        /// </summary>
+        /// <param name="collection">A list</param>
+        /// <param name="comparer">A comparer</param>
+        public BinarySearcher(IList<T> collection, Comparer<T> comparer)
+        {
+            if (collection == null)
+            {
+                throw new NullReferenceException("List is null");
+            }
+
+            this._collection = collection;
+            this._comparer = comparer;
+           
+        }
+
+     
+        /// <summary>
+        /// Apply Binary Search in a list.
+        /// </summary>
+        /// <param name="item">The item we search for</param>
+        /// <returns>If item found, its' index, -1 otherwise</returns>
+        public int BinarySearch(T item)
+        {
+            _currentItemIndex = -1;
+            _leftIndex = 0;
+            _rightIndex = _collection.Count - 1;
+            HeapSorter.HeapSort(_collection);
```

Comment:
Why don't you sort the `_collection` in the constructor? Any reason you're sorting here every time you perform a search operation?